### DISCUSSION TITLE
Patch updates for a dependency are skipped over in favor of minor updates when separateMinorPatch=false (default)

### DISCUSSION THREAD

**Discussion by mzeijen** [renovatebot/renovate#16488] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.107.1

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

GitLab Enterprise Edition 14.8.5-ee

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

_TLDR: When `separateMinorPatch` is set to `false` and minor updates are disabled then Renovate doesn't create any MR's for available patch updates within the same major/minor range, if a newer minor version is available. Having `separateMinorPatch` prevents Renovate from seeing the available patch update for the current minor version, which to me is unexpected ._

We have the following configuration (simplified, with only the relevant parts):
```json
{
  "enabledManagers": ["maven"],
  "separateMinorPatch": false,
  "packageRules": [
    {
      "description": "Disable anything but patch updates for all dependencies",
      "matchUpdateTypes": ["major", "minor"],
      "enabled": false
    }
  ]
}
```

Here you can see that we have `separateMinorPatch` turned to `false` (in reality we use the default `false` value and haven't set it explicitly), and we disabled major and minor updates.

With this configuration we expected to only get patch updates for our dependencies, but to our surprise we didn't get any patch updates for some of the dependencies even though patch updates where available within the same major/minor range. For example, one of the dependencies we expected to get a patch update was `org.springframework.cloud:spring-cloud-openfeign-dependencies`. We had it set to version `3.0.6` and we expected an MR for version `3.0.7` but that was never created. After a lot of troubleshooting we figured out that having `separateMinorPatch` set to `false` was preventing the creation of the patch MR's. When we set `separateMinorPatch` too `true` we started getting the `patch` updates. 

For me the `separateMinorPatch` setting created unexpected behaviour. I understand that in the situation where minor updates are not disabled, it should prevent any patch MR`s being created when a new minor update is available. But I didn't expect that it would also prevent patch updates when minor versions are disabled. What I expected is that when minor versions are disabled, Renovate will create patch updates within the same major/minor range, even when `separateMinorPatch` is set to `false`.

PS.
To be honest, I am not sure if this behaviour is a bug or you consider it a feature change. Because I didn't expect this behaviour I considered it a bug. 

### Relevant debug logs

[2022-07-08T12:05:35.490Z] DEBUG: renovate/39 on runner-bgp2l6xt-project-6288-concurrent-0r9vm9: packageFiles with updates (logContext=mI7goRc-1cZjTaMpjsdP0, repository=axle/axle-framework/axle)
    config: {
      "maven": [
        {
          "datasource": "maven",
          "packageFile": "axle-dependencies/pom.xml",
          "deps": [
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-commons-dependencies",
              "currentValue": "3.0.5",
              "fileReplacePosition": 967,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-commons.version",
              "depIndex": 0,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "3.1.3",
                  "newValue": "3.1.3",
                  "releaseTimestamp": "2022-05-26T17:36:58.000Z",
                  "newMajor": 3,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-commons-dependencies/",
              "currentVersion": "3.0.5",
              "isSingleVersion": true,
              "fixedVersion": "3.0.5"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-netflix-dependencies",
              "currentValue": "3.0.5",
              "fileReplacePosition": 813,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-netflix.version",
              "depIndex": 1,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "3.1.3",
                  "newValue": "3.1.3",
                  "releaseTimestamp": "2022-05-26T20:21:06.000Z",
                  "newMajor": 3,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-netflix-dependencies/",
              "currentVersion": "3.0.5",
              "isSingleVersion": true,
              "fixedVersion": "3.0.5"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-config-dependencies",
              "currentValue": "3.0.6",
              "fileReplacePosition": 662,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-config.version",
              "depIndex": 2,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "3.1.3",
                  "newValue": "3.1.3",
                  "releaseTimestamp": "2022-05-26T20:14:34.000Z",
                  "newMajor": 3,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-config-dependencies/",
              "currentVersion": "3.0.6",
              "isSingleVersion": true,
              "fixedVersion": "3.0.6"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-sleuth-dependencies",
              "currentValue": "3.0.5",
              "fileReplacePosition": 1041,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-sleuth.version",
              "depIndex": 3,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "3.1.3",
                  "newValue": "3.1.3",
                  "releaseTimestamp": "2022-05-26T21:02:51.000Z",
                  "newMajor": 3,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-sleuth",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-sleuth-dependencies/",
              "currentVersion": "3.0.5",
              "isSingleVersion": true,
              "fixedVersion": "3.0.5"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-contract-dependencies",
              "currentValue": "3.0.5",
              "fileReplacePosition": 737,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-contract.version",
              "depIndex": 4,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "3.1.3",
                  "newValue": "3.1.3",
                  "releaseTimestamp": "2022-05-27T00:44:06.000Z",
                  "newMajor": 3,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-contract-dependencies/",
              "currentVersion": "3.0.5",
              "isSingleVersion": true,
              "fixedVersion": "3.0.5"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-openfeign-dependencies",
              "currentValue": "3.0.6",
              "fileReplacePosition": 890,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-openfeign.version",
              "depIndex": 5,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "3.1.3",
                  "newValue": "3.1.3",
                  "releaseTimestamp": "2022-05-26T20:27:05.000Z",
                  "newMajor": 3,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-openfeign-dependencies/",
              "currentVersion": "3.0.6",
              "isSingleVersion": true,
              "fixedVersion": "3.0.6"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-kubernetes-dependencies",
              "currentValue": "2.0.5",
              "fileReplacePosition": 585,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-kubernetes.version",
              "depIndex": 6,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "2.1.3",
                  "newValue": "2.1.3",
                  "releaseTimestamp": "2022-05-27T01:57:30.000Z",
                  "newMajor": 2,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-kubernetes-dependencies/",
              "currentVersion": "2.0.5",
              "isSingleVersion": true,
              "fixedVersion": "2.0.5"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.cloud:spring-cloud-circuitbreaker-dependencies",
              "currentValue": "2.0.2",
              "fileReplacePosition": 1122,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "import",
              "groupName": "spring-cloud-circuitbreaker.version",
              "depIndex": 7,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "2.1.3",
                  "newValue": "2.1.3",
                  "releaseTimestamp": "2022-05-26T20:35:35.000Z",
                  "newMajor": 2,
                  "newMinor": 1,
                  "updateType": "minor",
                  "branchName": "renovate/spring-cloud"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-cloud/spring-cloud-build",
              "homepage": "https://projects.spring.io/spring-cloud/spring-cloud-circuitbreaker-dependencies/",
              "currentVersion": "2.0.2",
              "isSingleVersion": true,
              "fixedVersion": "2.0.2"
            }
          ],
          "parent": "pom.xml"
        },
        {
          "datasource": "maven",
          "packageFile": "pom.xml",
          "deps": [
            {
              "datasource": "maven",
              "depName": "com.bol.axle:axle",
              "currentValue": "7.15.0-SNAPSHOT",
              "fileReplacePosition": 1757,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "parent",
              "groupName": "revision",
              "editFile": "pom.xml",
              "depIndex": 0,
              "updates": [],
              "skipReason": "disabled"
            },
            {
              "datasource": "maven",
              "depName": "org.springframework.boot:spring-boot-dependencies",
              "currentValue": "2.5.14",
              "fileReplacePosition": 545,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "parent",
              "depIndex": 1,
              "updates": [
                {
                  "bucket": "non-major",
                  "newVersion": "2.7.1",
                  "newValue": "2.7.1",
                  "releaseTimestamp": "2022-06-23T12:38:20.000Z",
                  "newMajor": 2,
                  "newMinor": 7,
                  "updateType": "minor",
                  "branchName": "renovate/spring-boot"
                }
              ],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/spring-projects/spring-boot",
              "homepage": "https://spring.io/projects/spring-boot",
              "currentVersion": "2.5.14",
              "isSingleVersion": true,
              "fixedVersion": "2.5.14"
            },
            {
              "datasource": "maven",
              "depName": "com.bol.axle.maven.plugin:axle-pom-rewriter-maven-plugin",
              "currentValue": "5.0.0",
              "fileReplacePosition": 5318,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "build",
              "depIndex": 2,
              "updates": [],
              "warnings": [],
              "versioning": "maven",
              "homepage": "https://gitlab.tools.bol.com/axle/axle-pom-rewriter-maven-plugin",
              "currentVersion": "5.0.0",
              "fixedVersion": "5.0.0"
            },
            {
              "datasource": "maven",
              "depName": "org.apache.maven.plugins:maven-jar-plugin",
              "currentValue": "${maven-jar-plugin.version}",
              "fileReplacePosition": 5540,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "build",
              "skipReason": "version-placeholder",
              "depIndex": 3,
              "updates": []
            },
            {
              "datasource": "maven",
              "depName": "org.apache.maven.plugins:maven-javadoc-plugin",
              "currentValue": "${maven-javadoc-plugin.version}",
              "fileReplacePosition": 6921,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "build",
              "skipReason": "version-placeholder",
              "depIndex": 4,
              "updates": []
            },
            {
              "datasource": "maven",
              "depName": "org.codehaus.mojo:flatten-maven-plugin",
              "currentValue": "1.2.7",
              "fileReplacePosition": 2734,
              "registryUrls": [
                "https://repo.maven.apache.org/maven2",
                "https://artifactory.tools.bol.com/artifactory/maven-bol"
              ],
              "depType": "build",
              "groupName": "flatten-maven-plugin.version",
              "depIndex": 5,
              "updates": [],
              "warnings": [],
              "versioning": "maven",
              "sourceUrl": "https://github.com/mojohaus/flatten-maven-plugin",
              "currentVersion": "1.2.7",
              "fixedVersion": "1.2.7"
            }
          ],
          "parent": "pom.xml",
          "packageFileVersion": "${revision}"
        }
      ]
    }

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment, marked as answer]
that's by design and no bug

**Reply by breun** [reply]
Both minor and patch updates getting disabled when you only disable minor updates is by design? I also ran into this issue and this was also very unexpected to me.

**Comment by rarkins** [comment]
I don't understand the reason for separateMinorPatch=false. If you set that true, I would expect it to raise patch updates

**Reply by breun** [reply]
Yes, that's also what we've discovered, but

1. we did not know this setting existed, and that it is set to `false` by default.
2. when it is set to `false`, I would still expect a merge request for a patch update when I only explicitly disable minor updates.

**Reply by rarkins** [reply]
No, that's not how it works.

**Reply by breun** [reply]
I’ve noticed that, but can you explain why a user should expect that Renovate no longer creates merge requests for patch updates when the user disabled minor updates only?

**Reply by rarkins** [reply]
No, we don't have time to explain to every single user who disagrees with a design decision

**Reply by breun** [reply]
I understand that Renovate combines minor and patch updates into a single 'bucket' by default (`separateMinorPatch` set to `false` by default), but when minor updates are disabled as a user I would still expect any available patch updates to end up in that bucket and have Renovate create merge requests for those patch updates.

**Reply by rarkins** [reply]
The bucket contents are filled before the highest version within the bucket is selected, then the updateType is determined based on that selected version, and only after then is the rule applied to the derived update type. Your rule effectively says "Skip any PRs containing minor or major updates" and that's what happens. Your expectation seems to be that it also includes ".. but if any of those PRs could potentially have a patch available instead then go back a few steps, use the highest patch in the bucket instead, and continue" but it's too late in the process to do that - the bucket was already filled and the winning (highest) release selected. In theory we could *maybe* set separateMinorPatch implicitly if we see a minor update type being disabled in a package rule, but our package rules can be very complex and this is not necessarily easy to do.

This repo is an example of the default behavior: https://github.com/renovate-reproductions/ignore-minor/pulls?q=is%3Apr

Note how the left-pad PR was autoclosed because it's a minor update. Note that the chalk PR was not, because it was a patch update anyway. So the package rule does not mean "ALL patch updates are ignored when minor updates are disabled". What's actually happening is that by default "all patch update are ignored for all dependencies if a minor update is available instead".

**Reply by breun** [reply]
We are indeed understanding `separateMinorPatch: false` to mean something like "all patch update are ignored for all dependencies if a minor update is available instead". But with minor updates disabled we expected minor updates to not be considered at all (since they are disabled), so that all PRs for patch updates get created, regardless of whether a minor update is available or not. The fact that by default the creation of a PR depends on the availability of a minor update when you've explicitly disabled minor updates took us a very long time to discover. For us this caused PR's for seemingly random patch updates to not get created. Turns out it wasn't random, but the experience came pretty close. :)

If disabled update types would not make it into the bucket, or would be removed from the bucket before the effective updateType for the highest version gets determined, then I think the resulting behavior would make more sense. I think this is also what any Renovate user would want to happen. From a user standpoint I can't imagine that someone who disables minor updates only wants a patch update when there is no minor update available.

I hope that I've been able to show you why we think it would make more sense for Renovate to behave like we expected, but if you don't agree, then I guess we'll need to agree to disagree and we'll need to remember to set `separateMinorPatch` to `true` when we disable minor updates on a branch, because otherwise we will silently miss some patch updates on that branch.

**Reply by rarkins** [reply]
If you can find a way to do it elegantly which I'm overlooking, then you are welcome to submit a PR. It's not that we don't agree that it would be nice to handle it automatically, but instead that we think it would add too much complexity.

**Reply by breun** [reply]
It looks like the disabled types should be applied earlier in the process, so that they don't enter the bucket, or they should at least be removed again before the highest versions are determined. I can't judge how hard it would be to implement, because I'm not very familiar with JavaScript/TypeScript. 

From a user perspective this is what we experienced:

1. We wanted Renovate to only create PRs for patch updates on a branch.
2. We created a package rule like this: ```{
  "description": "Only create patch updates on the 7.x branch",
  "matchBaseBranches": ["7.x"],
  "matchUpdateTypes": ["major", "minor"],
  "enabled": false
}```
3. We saw Renovate create PRs for patch updates and no PRs for minor updates and concluded everything was fine.
4. A couple of weeks later we read a blog post about Spring Cloud 2020.0.6 being released and we wonder why Renovate did not create a PR to update `org.springframework.cloud:spring-cloud-dependencies` from `2020.0.5` to `2020.0.6`.
5. After a day of investigating we find out that setting `separateMinorPatch` to `true` for this branch is necessary to let Renovate create PRs for _all_ patch updates, and we suddenly get a whole bunch of PRs for other patch updates that were not created before (apparently because there were also minor updates available for those dependencies).

**Reply by Trainmaster** [reply]
Just ran into the same issue and agree with the author that this is unexpected and imho non-intuitive.
